File: Programmers/Lv3/부대복귀.py
from collections import deque

def solution(n, roads, sources, end):
    graph = [[] for _ in range(n + 1)]

    for road in roads:
        graph[road[0]].append(road[1])
        graph[road[1]].append(road[0])

    dist = [-1] * (n + 1)

    # 생각을 바꿔보자
    # destination -> start로 가도 된다.

    def bfs():
        q = deque()
        dist[end] = 0
        q.append(end)

        while q:
            now = q.popleft()

            for v in graph[now]:
                if dist[v] == -1:
                    dist[v] = dist[now] + 1
                    q.append(v)

    bfs()
    answer = []
    for source in sources:
        answer.append(dist[source])

    return answer

File: Programmers/Lv3/부대복귀.js
function solution(n, roads, sources, destination) {
  var answer = [];
  // start = destination
  var graph = Array.from(Array(n + 1), () => []);

  for (let [a, b] of roads) {
    graph[a].push(b);
    graph[b].push(a);
  }

  var start = destination;

  const bfs = (s) => {
    var q = [s];
    dist[s] = 0;

    while (q.length > 0) {
      var node = q.shift();

      for (let edj of graph[node]) {
        if (dist[edj] === -1) {
          dist[edj] = dist[node] + 1;
          q.push(edj);
        }
      }
    }
  };
  var dist = Array(n + 1).fill(-1);
  bfs(start);

  for (let end of sources) {
    answer.push(dist[end]);
  }

  return answer;
}
